File: app/src/main/java/com/example/myapplication/codingminutes_levelup/sorting_searching/QuickSort.java
package com.example.myapplication.codingminutes_levelup.sorting_searching;

import java.util.Arrays;

public class QuickSort {
    public static void main(String[] args) {
        int[] a = {10, 5, 2, 0, 7, 6, 4};
        quickSort(a, 0, a.length - 1);
        System.out.println("Result: " + Arrays.toString(a));
    }

    static int partition(int[] a, int s, int e) {
        int pivot = a[e];
        int i = s - 1;

        for (int j = s; j < e; j++) {
            if (a[j] < pivot) {
                i++;
                int temp = a[i];
                a[i] = a[j];
                a[j] = temp;
            }
        }
        int temp = a[i + 1];
        a[i + 1] = a[e];
        a[e] = temp;
        return i + 1;
    }

    static void quickSort(int[] a, int s, int e) {
        //base case
        if (s >= e) {
            return;
        }
        //recursive case
        int pi = partition(a, s, e);
        quickSort(a, s, pi - 1);
        quickSort(a, pi + 1, e);
    }
}

File: app/src/main/java/com/example/myapplication/algorithm/sorting/QuickSort.kt
package com.example.myapplication.algorithm.sorting




/**
 * QUICK SORT
 * ->Quick sort is like merge sort, like divide and conquer but using pivot concept
 * ->Quick sort is an algorithm to sort an array of numbers, when we can say sorted array is every number in that array should be in correct position.
 * ->We say element in correct position if every numbers which appears to left of that element is less than that element and every number which appears to right of that element is greater than that element, if this condition is apply to every element in array, we say array has been sorted.
 *
 * */
fun main() {
    val numbers = intArrayOf(7, 1, 3, 5, 2, 6, 4)
    val result = qs(numbers, 0, numbers.size - 1)
    println("LIST::${result.joinToString(",")}")
}


fun quickSort(numbers: IntArray, start: Int, end: Int): List<Int> {
    if (start < end) {
        val pivotIndex = partition(numbers, start, end)
        quickSort(numbers, start, pivotIndex - 1)
        quickSort(numbers, pivotIndex + 1, end)
    }

    return numbers.toList()
}

/**
 * Partition algorithm returns pivot index
 * */
fun partition(numbers: IntArray, start: Int, end: Int): Int {
    var pivotIndex: Int = start
    var pivot = numbers[end]
    for (i in start until end) {
        if (numbers[i] <= pivot) {
            //swap numbers[pivotIndex] with numbers[i]
            val temp = numbers[pivotIndex]
            numbers[pivotIndex] = numbers[i]
            numbers[i] = temp
            pivotIndex++
        }
    }
    //swap numbers[pivotIndex] with numbers[end]
    val temp = numbers[pivotIndex]
    numbers[pivotIndex] = numbers[end]
    numbers[end] = temp

    return pivotIndex
}

fun qs(array:IntArray,start:Int,end:Int):IntArray{
    if(start<end){
        var pivot = parti(array,start,end)
        qs(array,start,pivot-1)
        qs(array,pivot+1,end)
    }
    return array
}

fun parti(array:IntArray,start: Int,end: Int):Int{
    var pivot = array[end]
    var pi = start

    for(i in start until end-1){
        if(array[i]<=pivot){
            var temp = array[i]
            array[i] = array[pi]
            array[pi] = temp
            pi++
        }
    }

    var temp = array[pi]
    array[pi] = array[end]
    array[end] = temp
    return pi
}